lib/spack/spack/binary_distribution.py

```diff
@@ -852,117 +1216,98 @@ def extract_tarball(spec, filename, allow_root=False, unsigned=False,
+    Check if concrete spec exists on mirrors and return a list
+    indicating the mirrors on which it can be found
     """
-    global _cached_specs
+    if spec is None:
+        return []
 
     if not spack.mirror.MirrorCollection():
         tty.debug("No Spack mirrors are currently configured")
         return {}
 
-    for mirror in spack.mirror.MirrorCollection().values():
-        fetch_url_build_cache = url_util.join(
-            mirror.fetch_url, _build_cache_relative_path)
+    results = []
+    lenient = not full_hash_match
+    spec_full_hash = spec.full_hash()
 
-        tty.debug('Finding buildcaches at {0}'
-                  .format(url_util.format(fetch_url_build_cache)))
+    def filter_candidates(candidate_list):
+        filtered_candidates = []
+        for candidate in candidate_list:
+            candidate_full_hash = candidate['spec']._full_hash
+            if lenient or spec_full_hash == candidate_full_hash:
+                filtered_candidates.append(candidate)
+        return filtered_candidates
 
-        index_url = url_util.join(fetch_url_build_cache, 'index.json')
+    candidates = cache_manager.find_built_spec(spec)
+    if candidates:
+        results = filter_candidates(candidates)
 
-        try:
-            _, _, file_stream = web_util.read_from_url(
-                index_url, 'application/json')
-            index_object = codecs.getreader('utf-8')(file_stream).read()
-        except (URLError, web_util.SpackWebError) as url_err:
-            tty.debug('Failed to read index {0}'.format(index_url), url_err, 1)
-            # Continue on to the next mirror
-            continue
+    # Maybe we just didn't have the latest information from the mirror, so
+    # try to fetch directly.
+    if not results:
+        results = try_direct_fetch(spec,
+                                   force=force,
+                                   full_hash_match=full_hash_match)
+        if results:
+            cache_manager.update_spec(spec, results)
 
-        tmpdir = tempfile.mkdtemp()
-        index_file_path = os.path.join(tmpdir, 'index.json')
-        with open(index_file_path, 'w') as fd:
-            fd.write(index_object)
+    return results
 
-        db_root_dir = os.path.join(tmpdir, 'db_root')
-        db = spack_db.Database(None, db_dir=db_root_dir,
-                               enable_transaction_locking=False)
 
-        db._read_from_file(index_file_path)
-        spec_list = db.query_local(installed=False)
+def update_cache_and_get_specs():
+    """
+    Get all concrete specs for build caches available on configured mirrors.
+    Initialization of internal cache data structures is done as lazily as
+    possible, so this method will also attempt to initialize and update the
+    local index cache (essentially a no-op if it has been done already and
+    nothing has changed on the configured mirrors.)
+    """
+    cache_manager.update_local_index_cache()
+    cache_manager.regenerate_spec_cache()
```

Comment:
This works but can you make it so that `update_local_index_cache()` calls `regenerate_spec_cache()`? I think that users would always want to call these together, so we should combine them.